File: src/emulator/joypad.nim
import utils
import bitops
import sdl2

var
    keyState: uint8
    P1*: uint8

proc toggleBit(num: int): void = keyState.setBit(num)
proc resetBit(num: int): void = keyState.clearBit(num)

proc unsetKey*(key: cint): void =
    case key
    of K_e: resetBit(0) # ACTION: A
    of K_q: resetBit(1) # ACTION: B
    of K_LSHIFT: resetBit(2) # ACTION: SELECT
    of K_z: resetBit(3) # ACTION: START

    of K_d: resetBit(4) # DIRECTION: RIGHT
    of K_a: resetBit(5) # DIRECTION: LEFT
    of K_w: resetBit(6) # DIRECTION: UP
    of K_s: resetBit(7) # DIRECTION: DOWN
    else: return

proc setKey*(key: cint): void =
    case key
    of K_e: toggleBit(0) # ACTION: A
    of K_q: toggleBit(1) # ACTION: B
    of K_LSHIFT: toggleBit(2) # ACTION: SELECT
    of K_z: toggleBit(3) # ACTION: START

    of K_d: toggleBit(4) # DIRECTION: RIGHT
    of K_a: toggleBit(5) # DIRECTION: LEFT
    of K_w: toggleBit(6) # DIRECTION: UP
    of K_s: toggleBit(7) # DIRECTION: DOWN
    else: return

proc getP1State*(): (uint8, bool) =
    var 
        controlBits = P1.bitsliced(4..5)
        res: uint8 = 0xCF
        fallingEdge: bool

    case controlBits.uint2
    of 0b00: discard

    of 0b01:
        let keyIndex = firstSetBit(keyState.bitsliced(0..3))
        if keyIndex > 0: 
            let finalKeyIndex = keyIndex - 1
            fallingEdge = res.testBit(finalKeyIndex)
            res.clearBit(finalKeyIndex)

    of 0b10:
        let keyIndex = firstSetBit(keyState.bitsliced(4..7))
        if keyIndex > 0: 
            let finalKeyIndex = keyIndex - 1
            fallingEdge = res.testBit(finalKeyIndex)
            res.clearBit(finalKeyIndex)

    of 0b11: res = 0xFF

    return (res, fallingEdge)

File: src/joypad.c
#include "joypad.h"

#include <SDL2/SDL.h>
#include <stdint.h>

#include "utils.h"

JoypadState js = {0};

#define toggleBit(pos) setBit(js.keyState, pos);
#define resetBit(pos) clearBit(js.keyState, pos);

void unsetKey(int key) {
    switch (key) {
        case SDLK_z:
            resetBit(0);
            break;
        case SDLK_x:
            resetBit(1);
            break;
        case SDLK_RSHIFT:
            resetBit(2);
            break;
        case SDLK_RETURN:
            resetBit(3);
            break;

        case SDLK_RIGHT:
            resetBit(4);
            break;
        case SDLK_LEFT:
            resetBit(5);
            break;
        case SDLK_UP:
            resetBit(6);
            break;
        case SDLK_DOWN:
            resetBit(7);
            break;
        default:
            break;
    }
}

void setKey(int key) {
    switch (key) {
        case SDLK_z:
            toggleBit(0);
            break;
        case SDLK_x:
            toggleBit(1);
            break;
        case SDLK_RSHIFT:
            toggleBit(2);
            break;
        case SDLK_RETURN:
            toggleBit(3);
            break;

        case SDLK_RIGHT:
            toggleBit(4);
            break;
        case SDLK_LEFT:
            toggleBit(5);
            break;
        case SDLK_UP:
            toggleBit(6);
            break;
        case SDLK_DOWN:
            toggleBit(7);
            break;
        default:
            break;
    }
}

void setKeyMask() {
    uint8_t controlBits = BEXTR(js.P1, 4, 5);
    js.keyMask = 0xCF;
    uint8_t states = 0x00;

    switch (controlBits) {
        case 0b00: {
            states = 0xFF;
            break;
        }
        case 0b01: {
            states = BEXTR(js.keyState, 0, 3);
            break;
        }
        case 0b10: {
            states = BEXTR(js.keyState, 4, 7);
            break;
        }
    }
    js.keyMask &= ~states;

}
